server/src/storage/datastore.ts

```diff
@@ -64,35 +63,68 @@ const getAll = async (kind: string, filters?: Filter[]) => {
 };
 
 /**
- * Unpacks id of modified object from MutationResult object.
+ * A Datastore wrapper that inserts a particular entity with the specified Kind if another
+ * entity with the same value for the specified unique property does not already exist.
+ * An error is thrown if an entity with the same unique property value already exists.
+ * @param kind The Kind of the Entity
+ * @param entity The Entity to be added
+ * @param uniqueProperty The property that should be unique for the specified kind
  */
-const getIdFromMutationResult = (
-  mutationResult: google.datastore.v1.IMutationResult
-): ResponseId | undefined => mutationResult?.key?.path?.[0]?.id;
+const insertUniqueEntity = async (
+  kind: string,
+  entity: Entity,
+  uniqueProperty: Filter
+): Promise<Entity | undefined> => {
+  const transaction = datastore.transaction();
+  const query = transaction
+    .createQuery(kind)
+    .filter(uniqueProperty.property, uniqueProperty.value);
```

Comment:
Could you explain why this implementation is doing insertion and throwing error if item exists, as the function doc specifies? Looks like it's just a query definition to me.

Reply:
@xlai20 Yea this line is just creating the query. The insertion part is implemented in the try catch block from line 83 onwards

Reply:
Looks good then. The big block doc below gave me a wrong impression that that's another function definition.